Remove rusty hook and replace with optional cargo-husky hook

ezpuzz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
<!--- Separate Additions, Removals and Modifications -->
This removes rusty-hook which had some issues (see #2401 comment) with cargo-husky.  cargo-husky allows a user-defined git hook script which is more flexible than rusty-hook.  I've created a basic one which prompts the user to optionally lint their code (staged and unstaged changes at the moment) for warnings and errors that would prevent their changes from being merged.  cargo-husky will not overwrite hooks that user already has in their `.git/hooks` directory.

one note, cargo-husky fails silently if you have a custom `CARGO_TARGET_DIR` environment variable. see https://github.com/rhysd/cargo-husky/issues/24  it does not affect contributors ability to commit.

In the future we should consider a `git push` hook that runs `cargo test` optionally.
I also think `cargo clippy` is probably redundant with `cargo check` but `cargo check` is faster to find basic warnings...

We may also consider in the future using `RUSTFLAGS="-D warnings"` in CI to reduce warning accumulation.  We do not want to set this locally as it invalidates the users build cache (which results in very slow build) and exits upon the very first warning found.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
When I create a PR I am frustrated if it fails CI because of formatting or clippy errors.  I would like to be reminded to check them before committing.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I have run the hook multiple times locally using `git commit`.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.
- [ ] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Updated the content of the book if this PR would make the book outdated.
- [ ] I have added tests to cover my changes.
- [ ] My code is used in an example.

Blisto91 commented:
Just want to say that this is really cool! Did not know about this (or git hooks in general).

You are correct that running both cargo check and clippy is redundant since clippy also includes a cargo check. 
Tho cargo check is a lot faster by itself ofcourse.

ezpuzz commented:
@CleanCut has mentioned that it may be better to not use any hooks unless contributor somehow opts in.  Perhaps build.rs could prompt the user to enable the hooks?  cargo-husky installs them with its own build.rs.  There is also a cargo-husky environment variable `CARGO_HUSKY_DONT_INSTALL_HOOKS=true` that prevents it from installing the hook.

CleanCut reviewed (changes requested):
My vote would be to leave the git hook optional. Don't use cargo-husky or any other enforcement mechanism. Instead, place `pre-commit` file in `script/` and document in `CONTRIBUTORS.md` how you can copy it over to `.git/hooks/` if you want it to run automatically.

Though I see how having this hook in place could be very beneficial for frequent contributors with certain development workflows. I think it would still be a hindrance for a lot of others (even other frequent contributors).

Using cargo-husky still injects a git hook which will be run whether the user wants it to or not, and breaks external tooling (GUIs, IDEs, etc.) via the requirement to interact with the hook via stdin/stdout.  That's a big hurdle for a lot of people, compared to a very small task of copying in the hook for those who want it.

Note: There seem to be a bunch of unrelated `fmt`-like updates in this PR.

CleanCut commented:
> We may also consider in the future using `RUSTFLAGS="-D warnings"` in CI to reduce warning accumulation.

For sure! You just need to eliminate all the warnings first. 😄

ezpuzz commented on docs/CONTRIBUTING.md:
```diff
@@ -61,24 +61,21 @@ done the following things first:
-4. You added your change in docs/CHANGELOG.md and linked your pull request number.
-5. For new features or changes to an existing one,
-   add or change either the book tutorial or the examples.
+4. For new features or changes to an existing feature,
```
I've implemented Release Drafter to create a changelog automatically from merged Pull Requests and locked the master branch to only allow changes from PRs.  Release Drafter will create a pretty changelog for us based on the PRs merged and their labeling, we can even automate this with an autolabel bot based on which crate is changed, etc.

check out the legion draft release changelog here: https://github.com/amethyst/amethyst/releases every PR is now linked automatically and the authors of each PR are tagged.  When I first starting using amethyst I noticed multiple PRs that went stale after requesting the contributors to update the changelog or perform other manual tasks.  Even some where the changelog was updated by the contributor but then the markdown link wasn't correctly formatted and they were asked to fix it again.  This discourages contributors from submitting changes.

Release Drafter will make sure our changelog is complete and correct without us having to monitor it.

CleanCut replied:
👍🏼

ezpuzz commented on docs/CONTRIBUTING.md:
```diff
@@ -61,24 +61,21 @@ done the following things first:
-   * `cargo +nightly clippy --workspace --all-targets --features empty -Z unstable-options`
-   * `cargo test --all-targets --workspace --features empty`
-   * `cargo run --example $YOUR_EXAMPLE --features $YOUR_BACKEND`
+2. All of the following commands completed without warnings or errors.  You can pass whatever features are relevant for your change. CI will test with all features enabled.
```
I added a note about using the git hook to run these commands.  Keeping them up to date in at least three places (CI, here, git hook) is kind of painful.  If I make the git hook script a little nicer it run on demand, but you only really need it at pre-commit or pre-push time.

The extra indentation nests the bullets within the numbered item instead of being at the same level as the numbered item visually.  I use https://github.com/joeyespo/grip to edit markdown locally and check against github formatting.

ezpuzz commented:
> 
> 
> My vote would be to leave the git hook optional. Don't use cargo-husky or any other enforcement mechanism. Instead, place `pre-commit` file in `script/` and document in `CONTRIBUTORS.md` how you can copy it over to `.git/hooks/` if you want it to run automatically.
> 
> Though I see how having this hook in place could be very beneficial for frequent contributors with certain development workflows. I think it would still be a hindrance for a lot of others (even other frequent contributors).
> 
> Using cargo-husky still injects a git hook which will be run whether the user wants it to or not, and breaks external tooling (GUIs, IDEs, etc.) via the requirement to interact with the hook via stdin/stdout. That's a big hurdle for a lot of people, compared to a very small task of copying in the hook for those who want it.
> 
> Note: There seem to be a bunch of unrelated `fmt`-like updates in this PR.

I've moved the script and removed cargo-husky.  I'm sorry that I made a rash change that affected contributors negatively.  I was naively seeking to force contributors to lint their code before pushing.  I will seek additional community feedback about tooling changes like that in the future.  I had some negative feedback on changing the examples as well in #2579 

There was one unused dependency removed from linting and the two `extern crate amethyst` declarations actually broke the git hook sometimes by creating cyclical unstable dependency chains in cargo so I consider that change related.

ezpuzz commented:
> 
> 
> > We may also consider in the future using `RUSTFLAGS="-D warnings"` in CI to reduce warning accumulation.
> 
> For sure! You just need to eliminate all the warnings first. 😄
> 
> Update: I can't remember if we didn't add that yet because there were a bunch of existing warnings, or if we removed it because it kept breaking things due to warnings in the dependency chain at the time. It was one or the other. 🤷🏼‍♂️

we only have two left, one can probably be allowed but is bad function naming.  the other is a deprecation warning and i'm not sure of best practice there.  I think it's okay to merge with warnings when code is working, especially when trying to get a milestone release out.   Also leaving some warnings in master branch can give newcomers an easy positive contribution to make and get them involved.

A parallel not required `-D warnings` CI job could show broken status on PRs at least.

CleanCut commented:
> I've moved the script and removed cargo-husky. I'm sorry that I made a rash change that affected contributors negatively. I was naively seeking to force contributors to lint their code before pushing. I will seek additional community feedback about tooling changes like that in the future. I had some negative feedback on changing the examples as well in #2579

No worries! We try things. We learn. We adjust. It's all good 👍🏼  Hopefully the tone of my feedback is coming across as calm, thoughtful, and collaborative. That's my intent. 😄
